solrmarc_mixin/src/de/unituebingen/ub/ubtools/solrmarcMixin/TuelibMixin.java

```diff
@@ -996,4 +983,508 @@ public String getRecordingDate(final Record record) {
+
+        return formats;
+    }
+
+    /**
+     * Determine Mediatype For facets we need to differentiate between
+     * electronic and non-electronic resources
+     *
+     * @param record
+     *            the record
+     * @return mediatype of the record
+     */
+
+    public Set<String> getMediatype(final Record record) {
+        final Set<String> mediatype = new HashSet<>();
+        final Set<String> formats = getFormatIncludingElectronic(record);
+        final String electronicRessource = "Electronic";
+        final String nonElectronicRessource = "Non-Electronic";
+
+        if (formats.contains(electronicRessource)) {
+            mediatype.add(electronicRessource);
+        } else {
+            mediatype.add(nonElectronicRessource);
+        }
+
+        return mediatype;
+    }
+
+    /**
+     * Determine the publication date for "date ascending/descending" sorting in
+     * accordance with the rules stated in issue 227
+     *
+     * @param record
+     *            MARC record
+     * @return the publication date to be used for
+     */
+
+    public String getPublicationSortDate(final Record record) {
+
+        final Set<String> format = getFormatIncludingElectronic(record);
+
+        // Case 1 [Article or Review]
+        // Match also the case of publication date transgressing one year
+        // (Format YYYY/YY for older and Format YYYY/YYYY) for
+        // newer entries
+        if (format.contains("Article") || format.contains("Review")) {
+            final DataField _936Field = (DataField) record.getVariableField("936");
+            if (_936Field != null) {
+                final Subfield jSubfield = _936Field.getSubfield('j');
+                if (jSubfield != null) {
+                    final String yearOrYearRange = jSubfield.getData();
+                    return yearOrYearRange.length() > 4 ? yearOrYearRange.substring(0, 4) : yearOrYearRange;
+                } else {
+                    System.err.println("getPublicationSortDate [No matching subfield 'j' in field 936]: " + record.getControlNumber());
+                }
+            } else {
+
+                System.err.println("getPublicationSortDate [No matching 936 field:] " + record.getControlNumber());
+            }
+
+        // Case 2: Otherwise get from 260c 
+        } else {
+            final Set<String> dates = new LinkedHashSet<>();
+            final List<VariableField> list260 = record.getVariableFields("260");
+            for (final VariableField vf : list260) {
+                final DataField df = (DataField) vf;
+                final List<Subfield> currentDates = df.getSubfields('c');
+                for (final Subfield sf : currentDates) {
+                    // We have possibly a date that follows a different
+                    // calculation of times
+                    String currentDateStr = sf.getData();
+                    final String DIFFERENT_CALCULATION_OF_TIME_REGEX = ".*?\\[(.*?)\\=\\s*(\\d+)\\s*\\].*";
```

Comment:
Do we want to create a constant that never changes for every loop iteration?  Maybe the compiler will optimise this away but I wouldn't rely on it.